hashintel/hash
# Add preliminary Asana integration

## 🌟 What is the purpose of this PR?

This builds off the work in #1771 and adds an Asana integration. 

## 🔗 Related links





- [Asana task](https://app.asana.com/0/0/1203657039764068/f) _(internal)_

## 🚫 Blocked by

- [ ] ~~#1776~~ _(rebased to remove dependency instead)_

## 🔍 What does this change?

Please go commit by commit

## 📜 Does this require a change to the docs?

- Not yet, the task-executor isn't stable enough for usage

## ⚠️ Known issues

- You have to create the secrets directories yourself
- The catalog is copied into the docker image, rather than being provided at runtime through GraphQL or something. This means you need to generate the right `catalog.json` _before_ running the production deployment, which is fairly annoying. Even in the short term, ideally the catalog will be something that's provided as part of the request, where users can select the streams they want, rather than a physical file.
- The Airbyte Asana connector does not allow filtering by time range, you have to import *all* data within that stream every time you run it 
  - The app can't handle this well when you're trying to import the "stories" stream, which is multiple hundreds of thousands of entities

## 🐾 Next steps

N/A

## 🛡 What tests cover this?

None to my knowledge

## ❓ How to test this?

Similar to #1771 

1. Checkout the branch / view the deployment
1. Follow the README to start the app as normal (including `external-services` and `yarn dev`)
1. Run `yarn workspace @hashintel/hash-task-executor start`
1. Open the front-end and login to generate the appropriate cookies
1. Open the GraphQL playground (http://localhost:5001/graphql)
1. Acquire a [personal access token for Asana](https://app.asana.com/0/my-apps)
1. Try the following mutation
   ```graphql
   mutation {
     executeAsanaSpecTask
   }
   ```
1. Then you can try the following (replace the repositories, access token, and date as needed)
   ```graphql
   mutation {
     executeAsanaCheckTask(config:{
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Then you can run the following to download the "catalog" for the connector (a list of available streams)
   ```graphql
   mutation {
     executeAsanaDiscoverTask(config:{
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. The previous mutation should have created a file: `packages/hash/task-executor/src/tasks/source-asana/secrets/catalog.json`. Format the file using your IDE, and select a few "streams" that you want to download (delete the ones you do not want). The more you pick, the slower it will be to test. The "stories" and "tasks" streams will ingest a _**lot**_ of data.
1. Once you have modified `catalog.json` as you wish, run the following mutation (again updating parameters as you wish) 
   ```graphql
   mutation {
     executeAsanaReadTask(config:{
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Check the app, you should see new entity types, and going to the relevant entity type page should show entities available.

## 📹 Demo
